File: LeetCode/Python/33.search-in-rotated-sorted-array.py
# @before-stub-for-debug-begin
from python3problem33 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode id=33 lang=python3
#
# [33] Search in Rotated Sorted Array
#
from typing import List
# TODO: Solve (160 / 195)
"""
[3,1]
1
Ans = -1
Exp = 1
"""
# @lc code=start
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        k = 0
        for i in range(len(nums) - 1):
            if nums[i] > nums[i + 1]:
                k = i
                break
        
        l, r = 0, len(nums) - 1

        while l <= r:
            m = l + (r - l) // 2
            idx = m - k

            if nums[idx] > target:
                r = m - 1
            elif nums[idx] < target:
                l = m + 1
            else:
                if idx < 0:
                    idx += len(nums)
                print(idx)
                return idx

        return -1
# @lc code=end

File: LeetCode/CPP/33.search-in-rotated-sorted-array.cpp
// @before-stub-for-debug-begin
#include <vector>
#include <string>

using namespace std;
// @before-stub-for-debug-end

/*
 * @lc app=leetcode id=33 lang=cpp
 *
 * [33] Search in Rotated Sorted Array
 */
#include <bits/stdc++.h>
using namespace std;

// @lc code=start
class Solution
{
public:
    int search(vector<int> &nums, int target)
    {
        int pivot_index = 0;
        int low = 0, m, high = nums.size() - 1;

        if (nums[0] > nums[nums.size() - 1])
        {
            while (low < high)
            {
                m = low + (high - low) / 2;
                if (nums[m] < nums[0])
                {
                    high = m;
                }
                else
                {
                    low = m + 1;
                }
            }
            pivot_index = low;
        }
        if (target > nums[pivot_index])
        {
            low = pivot_index;
            high = nums.size() - 1;
        }
        else
        {
            low = 0;
            high = pivot_index;
        }

        while (low <= high)
        {
            m = low + (high - low) / 2;
            if (nums[m] < target)
            {
                low = m + 1;
            }
            else if (nums[m] > target)
            {
                high = m - 1;
            }
            else
            {
                return m;
            }
        }

        return -1;
    }
};
// @lc code=end
